File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int sum = 0;
        for(int weight: weights){
            sum += weight;
        }

        int low=0, high=sum;
        while(low<high){
            int mid=(low+high)/2;
            if(isPossible(weights,days,mid)){
                high = mid;
            }
            else{
                low = mid+1;
            }
        }

        return low;
    }

    private boolean isPossible(int[] weights,int days, int weightCapacity){
        int daysRequired = 0, weightSum = (int)1e9;
        for(int weight: weights){
            if(weight > weightCapacity){
                return false;
            }

            weightSum += weight;
            if(weightSum > weightCapacity){
                weightSum = weight;
                daysRequired++;
                if(daysRequired>days){
                    return false;
                }
            }
        }

        return true;
    }
}

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        auto [minIt, maxIt] = minmax_element(weights.begin(),weights.end());
        int min = *minIt, max = *maxIt;
        int low = max, high = (weights.size()*max);
        int ans = high;
        while(low<=high){
            int mid = (low+high)/2;
            if(isPossible(weights,days,mid)){
                ans = mid;
                high = mid-1;
            }
            else{
                low = mid+1;
            }
        }

        return ans;
    }

private:
    bool isPossible(vector<int>& weights, int reqDays, int certainCapacity){
        int daysTook = 1, currLoad = 0;
        for(int &wt: weights){
            currLoad += wt;
            if(currLoad > certainCapacity){
                daysTook++;
                currLoad = wt;
                if(daysTook > reqDays){
                    return false;
                }
            }
        }

        return true;
    }
};